libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp

```diff
@@ -0,0 +1,452 @@
+   constructor is not called until detect() returns true, so we
+   already know that we should setup the proximity sensor
+ */
+AP_Proximity_RPLidarA2::AP_Proximity_RPLidarA2(AP_Proximity &_frontend,
+                                               AP_Proximity::Proximity_State &_state,
+                                               AP_SerialManager &serial_manager) :
+                                               AP_Proximity_Backend(_frontend, _state)
+{
+    _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Lidar360, 0);
+    if (_uart != nullptr) {
+        _uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Lidar360, 0));
+    }
+    _cnt = 0 ;
+    _sync_error = 0 ;
+    _byte_count = 0;
+}
+
+// detect if a RPLidarA2 proximity sensor is connected by looking for a configured serial port
+bool AP_Proximity_RPLidarA2::detect(AP_SerialManager &serial_manager)
+{
+    return serial_manager.find_serial(AP_SerialManager::SerialProtocol_Lidar360, 0) != nullptr;
+}
+
+// update the _rp_state of the sensor
+void AP_Proximity_RPLidarA2::update(void)
+{
+    if (_uart == nullptr) {
+        return;
+    }
+
+    // initialise sensor if necessary
+    if (!_initialised) {
+        _initialised = initialise();    //returns true if everything initialized properly
+    }
+
+    // if LIDAR in known state
+    if (_initialised) {
+        get_readings();
+    }
+
+    // check for timeout and set health status
+    if ((_last_distance_received_ms == 0) || (AP_HAL::millis() - _last_distance_received_ms > COMM_ACTIVITY_TIMEOUT_MS)) {
+        set_status(AP_Proximity::Proximity_NoData);
+        Debug(1, "LIDAR NO DATA");
+    } else {
+        set_status(AP_Proximity::Proximity_Good);
+    }
+}
+
+// get maximum distance (in meters) of sensor
+float AP_Proximity_RPLidarA2::distance_max() const
+{
+    return 16.0f;  //16m max range RPLIDAR2, if you want to support the 8m version this is the only line to change
+}
+
+// get minimum distance (in meters) of sensor
+float AP_Proximity_RPLidarA2::distance_min() const
+{
+    return 0.20f;  //20cm min range RPLIDAR2
+}
+
+bool AP_Proximity_RPLidarA2::initialise()
+{
+    // initialise sectors
+    if (!_sector_initialised) {
+        init_sectors();
+        return false;
+    }
+    if (!_initialised) {
+        reset_rplidar();            // set to a known state
+        Debug(1, "LIDAR initialised");
+        return true;
+    }
+
+    return true;
+}
+
+void AP_Proximity_RPLidarA2::reset_rplidar()
+{
+    if (_uart == nullptr) {
+        return;
+    }
+    uint8_t tx_buffer[2] = {RPLIDAR_PREAMBLE, RPLIDAR_CMD_RESET};
+    _uart->write(tx_buffer, 2);
+    _resetted = true;                   ///< be aware of extra 63 bytes coming after reset containing FW information
+    Debug(1, "LIDAR reset");
+    hal.scheduler->delay(RESET_RPA2_WAIT_MS);                      ///< wait at least 8ms after sending new request
```

Comment:
Any Ideas on How to improve and or avoid this hal.scheduler->delay(RESET_RPA2_WAIT_MS); call ?